data-products/src/main/scala/org/sunbird/analytics/job/report/CourseMetricsJob.scala

```diff
@@ -141,88 +137,33 @@ object CourseMetricsJob extends optional.Application with IJob with ReportGenera
-      .select(col("provider"), col("idtype"), col("externalid"), col("userid")).persist()
-
-    val custRootOrgId = getCustodianOrgId(loadData)
-    val detailsByUser = getUserSelfDeclaredDetails(userDF, custRootOrgId, externalIdentityDF, locationDF)
-    val detailsByState = getStateDeclaredDetails(userDF, custRootOrgId, externalIdentityDF, organisationDF, userOrgDF, locationDF)
-    
-    val userInfoDF = detailsByUser.union(detailsByState)
-
-    /*
-    * userDenormDF lacks organisation details, here we are mapping each users to get the organisationids
-    * */
-    val userRootOrgDF = userDF
-      .join(userOrgDF, userOrgDF.col("userid") === userDF.col("userid") && userOrgDF.col("organisationid") === userDF.col("rootorgid"))
-      .select(userDF.col("*"), col("organisationid"))
-
-    val userSubOrgDF = userDF
-      .join(userOrgDF, userOrgDF.col("userid") === userDF.col("userid") && userOrgDF.col("organisationid") =!= userDF.col("rootorgid"))
-      .select(userDF.col("*"), col("organisationid"))
-
-    val rootOnlyOrgDF = userRootOrgDF
-      .join(userSubOrgDF, Seq("userid"), "leftanti")
-      .select(userRootOrgDF.col("*"))
-
-    val userOrgDenormDF = rootOnlyOrgDF.union(userSubOrgDF)
-
-    val locationDenormDF = userOrgDenormDF
-      .withColumn("exploded_location", explode(col("locationids")))
-      .join(locationDF, col("exploded_location") === locationDF.col("id") && locationDF.col("type") === "district")
-      .select(col("name").as("district_name"), col("userid"))
-
-    /**
-      * Resolve the block name by filtering location type = "BLOCK" for the locationids
-      */
-    val blockDenormDF = userOrgDenormDF
-      .withColumn("exploded_location", explode(col("locationids")))
-      .join(locationDF, col("exploded_location") === locationDF.col("id") && locationDF.col("type") === "block")
-      .select(col("name").as("block_name"), col("userid"))
-
-    val userLocationResolvedDF = userOrgDenormDF
-      .join(locationDenormDF, Seq("userid"), "left_outer")
-
-    val userBlockResolvedDF = userLocationResolvedDF.join(blockDenormDF, Seq("userid"), "left_outer")
-
-    /*
-    * Resolve organisation name from `rootorgid`
-    * */
-
-    val resolvedOrgNameDF = userBlockResolvedDF
-      .join(organisationDF, organisationDF.col("id") === userBlockResolvedDF.col("rootorgid"), "left_outer")
-      .select(userBlockResolvedDF.col("userid"), userBlockResolvedDF.col("rootorgid"), col("orgname").as("orgname_resolved"))
+  def getUserData()(implicit spark: SparkSession): DataFrame = {
+    spark.read.format("org.apache.spark.sql.redis")
+      .option("keys.pattern", "*")
+      .option("infer.schema", true).load()
+  }
 
-    userBlockResolvedDF
-      .join(userInfoDF, Seq("userid"), "left_outer")
-      .join(resolvedOrgNameDF, Seq("userid", "rootorgid"), "left_outer")
-      .dropDuplicates("userid")
-      .persist()
+  def getTPDBatches(config: JobConfig): List[CourseInfo] = {
```

Comment:
It should not be only for "TPD" it should handle other batch types as well. since your method name says only TPD and pass required config to this method instead of passing whole config.